### PR TITLE
Tags migration

### PR DESCRIPTION
**Pull request by afreydev**
# Summary

This PR migrate the snapshot tags field to a new independent model.

# Related #425 

# Changes these areas

- [ ] Bugfixes
- [x] Feature behavior
- [ ] Command line interface
- [ ] Configuration options
- [ ] Internal architecture
- [ ] Archived data layout on disk

### CODE REVIEW

File: archivebox/core/admin.py [hunk 1]
```diff
@@ -66,6 +66,12 @@ class SnapshotAdmin(admin.ModelAdmin):
     actions = [delete_snapshots, overwrite_snapshots, update_snapshots, update_titles, verify_snapshots]
     actions_template = 'admin/actions_as_select.html'
 
+    def get_queryset(self, request):
+        return super().get_queryset(request).prefetch_related('tags')
+
+    def tag_list(self, obj):
+        return u", ".join(o.name for o in obj.tags.all())
```
Comment (pirate) [inline archivebox/core/admin.py:73]:
```suggestion
        return ', '.join(obj.tags.values_list('name', flat=True))
```

(In python 3 all strings are `u""` strings by default.)

**Comment by cdvv7788** [comment]
LGTM. I fixed a couple issues that were still present. It seems to work pretty well. @pirate can you review it please?

**Review by cdvv7788 — APPROVED** [review]

File: archivebox/core/models.py [hunk 2]
```diff
@@ -50,6 +60,13 @@ def as_link(self) -> Link:
     def as_link_with_details(self) -> Link:
         from ..index import load_link_details
         return load_link_details(self.as_link())
+    
+    def get_tags_str(self) -> str:
+        tags = ','.join(
+            tag.name
+            for tag in self.tags.all()
+        ) if self.tags.all() else ''
```
Comment (pirate) [inline archivebox/core/models.py:68]:
```suggestion
        tags = ','.join(self.tags.values_list('name', flat=True)) if self.tags.exists() else ''
```
minor change, this is a migration so you can ignore it, but for future code this pattern is faster ^

**Review by pirate — APPROVED** [review]
Why was the decision made to use `django-taggit` instead of making our own `Tag` model?

I have a mild preference for making our own over using a library, but I assume there was some good reasoning for going the library route.

**Comment by cdvv7788** [comment]
Well, we can definitely change it. A ForeignKey will be better than a GenericRelationship. The base is there, the refactor should not be complex. If you prefer it that way, we can just change it. Given that tags won't be used in any other model, it makes sense.
Taggit has several defaults that make it faster to implement, but other than that there is not a strong reason to keep it around.

Should we change it then?

**Comment by pirate** [comment]
I think I'd like to change it then, as I much prefer the `ForeignKey` to `GenericRelationship`. Sorry for the trouble.

**Comment by cdvv7788** [comment]
No worries. The base is already there, so the refactor should not be complicated.

**Comment by cdvv7788** [comment]
@pirate this requires a m2m relationship, not a ForeignKey. I will still refactor it, but wanted to correct that point.

**Comment by cdvv7788** [comment]
@pirate Please review again. There are some rough edges, but it works well enough.

File: archivebox/core/models.py [hunk 3]
```diff
@@ -50,6 +92,13 @@ def as_link(self) -> Link:
...
```
Comment (pirate) [inline archivebox/core/models.py:100]: return ','.join(self.tags.values_list('name', flat=True))

**Pull request merged**